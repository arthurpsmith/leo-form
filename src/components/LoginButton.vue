<template>
  <span>
    <v-btn
      v-if="!isLoggedIn"
      color="primary"
      class="ml-4"
      @click="attemptLogin"
    >{{ buttonText }}</v-btn>
    <v-menu
      v-if="isLoggedIn"
      class="ml-4"
    >
      <template v-slot:activator="{ on, attrs }">
        <v-btn
          outlined
          v-bind="attrs"
          v-on="on"
          class="ml-4"
        >
          {{ userName }}
          <v-icon>mdi-menu-down</v-icon>
        </v-btn>
      </template>
      <v-list>
        <v-list-item @click="authMachine.send('LOG_OUT')">
          <v-list-item-title>{{ buttonText }}</v-list-item-title>
        </v-list-item>
      </v-list>
    </v-menu>
</span>
</template>

<script lang="ts">
import { defineComponent } from '@vue/composition-api'
import { useAuthService } from '@/statemachines/auth.machine'
import { StateValue } from 'xstate'

export default defineComponent({
  name: 'LoginButton',
  setup () {
    const authMachine = useAuthService()
    const attemptLogin = () => {
      authMachine.send({
        /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
        // @ts-ignore
        type: 'ATTEMPT_LOG_IN',
        /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
        // @ts-ignore
        userDetails: {
          /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
          // @ts-ignore
          username: 'someone@crossref.org'
        }
      })
    }
    return {
      authMachine, attemptLogin
    }
  },
  computed: {
    authState (): StateValue {
      return this.authMachine.state.value.value
    },
    userName (): string {
      return (typeof this.authMachine.state.value?.context?.userDetails !== 'undefined') ? this.authMachine.state.value.context.userDetails.username : ''
    },
    buttonText (): string {
      switch (this.authState) {
        case 'loggedIn':
          return 'Log out'
        default:
          return 'Log in'
      }
    },
    isLoggedIn (): boolean {
      return this.authState === 'loggedIn'
    }
  }
})
</script>
