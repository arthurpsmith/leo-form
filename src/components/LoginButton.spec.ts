import { createLocalVue, mount } from '@vue/test-utils'
import LoginButton from '@/components/LoginButton.vue'
import Vuetify from 'vuetify'
import CompositionApi from '@vue/composition-api'
import { getAuthService, authSymbol } from '@/statemachines/auth.machine'
import { useActor } from 'xstate-vue2'

describe('LoginButton.vue', () => {
  const localVue = createLocalVue()
  localVue.use(CompositionApi)
  let vuetify: Vuetify
  let authServiceToProvide: ReturnType<typeof getAuthService>
  let provideWithExportedSymbol: Record<string, unknown>
  let authService
  function mountFunction (options = {}) {
    return mount(LoginButton, {
      localVue,
      vuetify,
      provide: provideWithExportedSymbol,
      ...options
    })
  }
  beforeEach(() => {
    vuetify = new Vuetify()
    authServiceToProvide = getAuthService()
    provideWithExportedSymbol = {
      [authSymbol as symbol]: authServiceToProvide
    }
    authService = useActor(authServiceToProvide)
  })
  it('should work', () => {
    const wrapper = mountFunction()

    expect(wrapper.html()).toMatchSnapshot()
  })
  it('Updates button text on login success', async () => {
    const wrapper = mountFunction()
    const button = wrapper.find('button')
    await button.trigger('click')
  })
  // Probably a better candidate for an e2e test
  // it('Calls the changeLocale method with the new locale on menu item click', async () => {
  //   const wrapper = mountFunction()
  //   await wrapper.find('button').trigger('click')
  //   const menuItems = wrapper.findAll('.v-list-item')
  //   const newLocateItem = menuItems.at(2)
  //   await newLocateItem.trigger('click')
  //   expect(spy).toBeCalledTimes(1)
  //   expect(menuItems).toHaveLength(3)
  // })
})
