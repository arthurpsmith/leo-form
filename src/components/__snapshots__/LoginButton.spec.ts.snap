// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginButton.vue should work 1`] = `
<span><button type="button" class="ml-4 v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary"><span class="v-btn__content">Log in</span></button>
<!----></span>
`;
